Extend CDROM provider to support any common type of disk/partition with ISO format

errordeveloper opened the pull request.

rn commented:
The diff in GH and git does not look right. Can you do a `git mv provider_cdrom.go provider_iso.go` and then make the changes. That way one actually sees the differences. Or even split this up into two commits. One which does the rename and one which adds the new functionality? That way one can see the actual changes.

It would also be good if the commit message explains more of the rationale for this change.

ijc commented:
`git mv` is no different to `git rm && git add` in terms of what is stored in the repo, there is no specific metadata recorded about such things. The "move" and "copy" display is entirely a presentation layer thing. IOW `git mv` wouldn't help here -- it's just that the old and new files differ enough that GH didn't guess there was a move involved while rendering the diff.

For the CLI you can control the heuristics a bit with the `-M` and `-C` options to `git show`, I suppose that isn't possible in the GH webui though.

Making it two commits would indeed be one way to avoid the problem though. and +1 to better commit messages...

errordeveloper commented:
Sounds good, will do! I have indeed use 'git mv'.

On Wed, 30 Jan 2019, 10:45 am Ian Campbell <notifications@github.com wrote:

> git mv is no different to git rm && git add in terms of what is stored in
> the repo, there is no specific metadata recorded about such things. The
> "move" and "copy" display is entirely a presentation layer thing. IOW git
> mv wouldn't help here -- it's just that the old and new files differ
> enough that GH didn't guess there was a move involved while rendering the
> diff.
>
> For the CLI you can control the heuristics a bit with the -M and -C
> options to git show, I suppose that isn't possible in the GH webui though.
>
> Making it two commits would indeed be one way to avoid the problem though.
> and +1 to better commit messages...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/linuxkit/linuxkit/pull/3280#issuecomment-458898297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS6H_48NUd_lBorYFRVLZVLQA9oHTks5vIXe3gaJpZM4aZMe->
> .
>

errordeveloper committed:
Generalise CDROM provider to support common block devices

CDROM maybe convenient for virtual machines, yet with bare-meta
it actually implies that machine has to have an optical dirve.
It's much more covenient to use USB flash drives instead, this
change allows for any SCSI disk devices and MMC block devices,
as well as adds support for SCSI CDROMs with new prefix.
This change simply expands the range of supported block devices,
the provider implementation continues to rely on ISO filesystem.

Simply put - you can now put metadata.iso onto a separate partition
of a USB drive or SD card, and it will be recognised by the
metadata package.

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
Rename provider_cdrom.go -> provider_iso.go

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "metadata-iso-disk" git@github.com:errordeveloper/linuxkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353890016
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

errordeveloper commented:
I've split the commits as advised.

rn commented:
Thanks for the update. the diff is not much easier to read.

I'm not sure I like this approach. This seems to mount/unmount all disk type devices instead of stopping when one suitable device is found.

errordeveloper commented:
> the diff is not much easier to read.

I thought https://github.com/linuxkit/linuxkit/pull/3280/commits/5c3bde1f775ac17feb658db0abb93531050141ec fairly trivial...

> I'm not sure I like this approach. This seems to mount/unmount all disk type devices instead of stopping when one suitable device is found.

That's just what it did before, only limited to all `/dev/sr*` devices. If we believe that behaviour should change, I'm more than happy to implement that in this PR, or perhaps separate PR would make sense.
Perhaps the current behaviour was somewhat satisfactory, considering that systems with many optical drives aren't very common. I agree that trying to mount each and every drive is not a good idea, perhaps pre-filtering on filesystem type would be a good idea? We could even match on a well-known ISO volume ID, e.g. something like 'LinuxKit metadata'.

rn commented:
Sorry, I meant that the diff is *now* much easier to read.

But the code is mounting every disk in the system. And only later, during probe, it goes through the providers and stops on the first suitable on. This doesn't look nice. Why not mount only during probe and stop mounting once a suitable device is found. I dn't think we want to make it more complicated with filesystem probing in whatnot, just stop mounting once you found a suitable device.

errordeveloper commented:
Sure, that makes sense. Happy to adjust the behaviour that way.

errordeveloper committed:
Improve ISO provider

- mount one devices at a time
- stop after first suitable device is found
- unmount and remove mount point properly
- prioritize MMC devices over any SCSI disks

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented:
@rn please take a look at 247bdaf.

errordeveloper commented:
There is a challenge with USB devices, I ended up having to do this:

```yaml
onboot:
  - name: modprobe-usb-storage
    image: linuxkit/modprobe:v0.6
    command: ["modprobe", "usb-storage"]
  - name: wait-usb-storage
    image: linuxkit/alpine:3683c9a66cd4da40bd7d6c7da599b2dcd738b559
    binds:
      - /dev:/dev
    command:
      - sh
      - -c
      - until test -b /dev/sdc ; do sleep 1 ; done
  - name: metadata
    image: errordeveloper/metadata:9250ccfc8a4ba879989281b5f782bae2551521b9
```

I might find a better way, but this is what I have to do at the moment.

errordeveloper commented:
@rn would you mind to take another look at this please?
